rfcs/032-calm-deletions/README.md

```diff
@@ -0,0 +1,30 @@
+# RFC 032: Calm deletion watcher
+
+**Status:** Draft
+
+**Last updated:** 28/01/2021
+
+## Background
+
+We've realised that the pipeline does not pick up deleted Calm records - once they're deleted there is no trace of them in the Calm API, but they remain in the source VHS tables. Especially because a cleanup project at the end of last year was responsible for a very large number of deletions, this needs to be resolved.
+
+So far as we can tell, there is no way to find deleted records other than checking for their absence. This means that our solution needs to take some sort of polling-based approach: going through the source table and checking for the continued presence of each record in the Calm API.
+
+## Proposal
+
+- The deletion watcher will be an ECS task within the existing calm adapter cluster.
+- It will run as a scheduled task at a frequency to be agreed with the collections information team.
+- It will also be possible to run the task manually via the `RunTask` API. A script will be provided for this.
+- The task will scan the source data VHS table and check for the existence of each record in the Calm API. If the record has been deleted (is no longer present) then it will update the VHS entry to flag the deletion. 
+- The calm transformer will check for the presence of the deletion flag in the source data and create `Deleted` works as appropriate.
+
+This architecture is illustrated below (arrows indicate the direction of data flow):
+
+![image](https://user-images.githubusercontent.com/4429247/106134857-08103780-615f-11eb-845f-112109ead907.png)
+
+## Questions & potential issues
+
+- Are we really 100% sure that polling for deletions is the only way to detect them?
```

Comment:
It's one of (at least) two approaches, in my mind.

You could also reharvest the entire Calm catalogue into a fresh VHS, compare the IDs, and work out what's missing.

There are ~370k items in the current Calm VHS. Call it 400k to keep the numbers simple. Now suppose you have two copies of it, in memory, that's 800k. If each UUID takes a kilobyte of memory, that's 800MB, which should fit into the memory of one of our apps. Diff those two sets, voila, you have your deleted items.

The optimal approach would probably be a hybrid of both:

* Automatically and continually polling records that haven't changed in a while on a fixed schedule, and pruning them that way
* A button that lets us do a complete reharvest and mark any missing records as deleted

Reply:
Not sure I understand - what's the advantage of having an intermediate VHS for comparison? It seems like it would be much slower and more expensive.

I agree that we need both of those bullet points, I think this RFC _mostly_ covers them but seeing whether they've recently been changed needs a whole extra trip to the S3 portion of the VHS - is it worth it?

Reply:
Yeah, you don't really need the intermediate VHS unless you plan to drop the existing one (which we don't, but we should write down why it's not enough to reharvest and move on).

That depends on what data we record in the Calm DynamoDB table, and whether we want to record freshness checks somewhere. I idly wondered whether we'd put a `lastFetchedFromCalm` field in DynamoDB which the freshness checker would update, but I think that's probably wrong – a whole pile of conditional writes we mostly don't care about.

Reply:
It occurs to me as well that a reharvest may also be a little tricksy, because it takes a while.

If there's a record in the Calm VHS that you didn't see during the reharvest, is it because that record got deleted before you started the reharvest, or added afterwards?

Reply:
I don't really follow this idea of "freshness" - could you elaborate on that? All we're checking for is whether things have been deleted.

Reply:
Ah sorry.

For me, the notion of "freshness" is about _"when did I last _**know**_ this thing was in Calm"_?

If a record was updated yesterday, it's very "fresh" – it seems unlikely that it's been deleted.<br/>
If a record was updated last year and we haven't polled for it in a month, it's more likely that it's been deleted.

And I was wondering if that might affect how the deletion checker works, e.g. _"I don't re-poll for a record that was updated less than a week ago"_ (depending on how C&I feel).

Reply:
Yeah, I did wonder about these optimisations but I think they're really quite complicated for minimal gain - record freshness is going to follow a power law distribution where the _vast_ majority of records haven't been updated in a long time, so the reduction in requests that doing this would offer is likely to be a fairly small percentage of the total.